platforms/jvm/toolchains-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/DefaultJavaToolchainProvisioningService.java

```diff
@@ -185,4 +209,77 @@ public BuildOperationDescriptor.Builder description() {
                 .progressDisplayName(displayName);
         }
     }
+
+    private static class ToolchainDownloadFailureTracker {
+
+        private final Map<String, Exception> resolveFailures = new TreeMap<>();
+        private final Map<String, Exception> provisioningFailures = new TreeMap<>();
+
+        public void addResolveFailure(String repositoryName, Exception failure) {
+            resolveFailures.put(repositoryName, failure);
+        }
+
+        public void addProvisioningFailure(String repositoryName, Exception failure) {
+            provisioningFailures.put(repositoryName, failure);
+        }
+
+        public ToolchainDownloadFailedException buildFailureException() {
+            String message = "No locally installed toolchains match and the configured toolchain download repositories aren't able to provide a match either." +
+                (hasFailures() ? " " + failureMessage() : "");
+
+            String[] resolutions = {
+                "Learn more about toolchain auto-detection at " + Documentation.userManual("toolchains", "sec:auto_detection").getUrl() + ".",
+                "Learn more about toolchain repositories at " + Documentation.userManual("toolchains", "sub:download_repositories").getUrl() + "."
+            };
+
+            ToolchainDownloadFailedException exception = new ToolchainDownloadFailedException(message, resolutions);
+
+            return addFailuresAsSuppressed(exception);
+        }
+
+        private <T extends Exception> T addFailuresAsSuppressed(T exception) {
+            for (Exception resolveFailure : resolveFailures.values()) {
+                exception.addSuppressed(resolveFailure);
+            }
+
+            for (Exception provisionFailure : provisioningFailures.values()) {
+                exception.addSuppressed(provisionFailure);
+            }
+
+            return exception;
+        }
+
+        public void logFailuresIfAny() {
+            if (hasFailures()) {
+                LOGGER.warn(failureMessage() + ". Switch logging level to DEBUG (--debug) and look at the stacktrace (--stacktrace) for further information.");
```

Comment:
🤔 I'm not sure `--stacktrace` is relevant here given we simply log the exception without throwing it